### ISSUE TITLE
[Gameplay/Camera] Account for characters leaving the camera's view

### ISSUE THREAD

**Issue by ribbybibby** [ribbybibby/Grimm-Journey#56]
Now that we have camera panning (#49), we have to implement a way to deal with characters straying off screen or too far towards the edges of the camera's view. 

I think we have all agreed that there should be some negatives to this:
- Swirling darkness
- Bad aura that grows
- Damage to both characters
- Shaky camera
- Evil audio

And possibly some UI:
- An arrow pointing at the offending character's position off-screen

Programatically I have a couple of thoughts on how I can easily trigger these events:
1. Trigger bad JuJu when the distance between the characters is >= to 85-95% of the screen height (use orthographic size to work out camera dimensions) 
2. Trigger bad JuJu when character actually leaves the screen limits (again, camera dimensions to work out edge co-ords) 
3. ...and then trigger a timer event which increments the evil stuff until the character returns or the distance between the two lessens

**Comment by ribbybibby** [comment]
I'm going to try implementing some placeholder mechanics into my camera in preparation for decisions being made on where we go with this. I'm thinking:

**A way of detecting that either character is offscreen:**

```C#
float topY = gameObject.transform.position.y + gameObject.camera.orthographicSize;
float bottomY = gameObject.transform.position.y - gameObject.camera.orthographic.Size;
bool offScreen;

if (bbw.transform.position.y >= topY || bbw.transform.position.y <= bottomY)
{
    Debug.Log ("BBW is offscreen, run code!");
    offScreen = true; 
}
else
{
    Debug.Log ("BBW is onscreen, don't run code!");
    offScreen = false;
}
```

**A way of incrementing a number (for some use somehow) for every update that the character is offscreen:**

```C#
int someNumber = 0;

if (offScreen == true) 
{
    someNumber++;
}
if (offScreen = false)
{
    someNumber = 0;
}
```
(This could be used to control the size, speed or ferocity of the swirling darkness/doomy aura)

**I also want to look at how to code a camera shake:**
Not as sure on how to achieve this. My initial attempt would be to generate a tiny range around the camera's centre position and then transform to this based on a RNG every update.

**Comment by ribbybibby** [comment]
In bcb0ed04ba5103df977830ab8b318cda98c17080, I added the two bits of code from above into CameraMove.cs as the method Offscreen(). Now to start on the camera shake...

**Comment by HoboMechanistic** [comment]
When coding the shake in Update, you'd have to make sure you don't use a completely RND solution. I tried this with the witch ages ago and she would shake rapidly all over the place, up down and around as if she was having a siezure. You might have better luck moving the camera in a direction over time?

**Comment by ribbybibby** [comment]
I've added a fairly simple camera shake in 18f74c0dbacb7b62db210cd82195c30c93a8811a that triggers when one (or both) of the characters leave the camera's view. I managed to skirt around the issues with RND that Rob2 mentioned by returning to the centre every other update, which seems to create a fairly even effect.

The ferocity of the shake can be adjusted via the public variable shakeForce (0.3 seems like a good number for now).

One improvement I'd like to make is to tie the force of the shake to the incrementing number from OffScreen(); this will require me to consider the maths of converting the someNumber value to the shakeForce value without making the shake ridiculous. 

Lastly, please note that I have only added this shake (and the OffScreen() stuff) to Cameras 4 and 5 for the moment, just while these features are kind of up in the air.

**Comment by ribbybibby** [comment]
I've removed shakeForce and now use someNumber to control the ferocity of the shake. This means that it intensifies and subsides based on how long the char is offscreen or onscreen for. 

I can't commit it at the moment, because I'm working on a coach back to London but it'll make it in tomorrow!

**Comment by ribbybibby** [comment]
To do:
- Devise and add solution for swirling darkness/bad aura
- Code in a fail state where the game will end if a character is off screen for too long

**Comment by ribbybibby** [comment]
Added improved camera shake in 232f62841d4712905f5425772a4c545f0d4a8bc1

**Comment by ribbybibby** [comment]
Assigning to myself. Going to add the game over fail state and then the swirling darkness.

**Comment by ribbybibby** [comment]
Closing as Fixed in 0c114873e8b7906159971c22e54537e035438070

We don't have a 'swirling darkness' animation but what we do have is a couple of black spots that will fade in and out depending on whether the char is off screen or not. In combination with the shake, this creates an ominous enough doom for the time being.

**Comment by ribbybibby** [comment]
Oh,  I also added a game over state when the char has been offscreen for too long.
